nimbus_nlp/question_classifier.py

```diff
@@ -1,287 +1,134 @@
 
-        # print("Predicted")
-        # print(predicted_tags)
-        # print()
+            # Add [VARIABLES] with weight 90.
+            # token.i returns the index of the token, and token.nbor(n) return the token
+            # n places away. Only the left neighbor is tested for brevity.
+            elif token.i != 0 and token.nbor(-1).text == "[":
+                features[token.text] = 90
 
-        # Compares all WH words in the tags array and returns False if one doesn't match
-        min_tag_len = min(len(test_tags), len(predicted_tags))
-        wh_match = True
-        i = 0
-        while (wh_match and i < min_tag_len):
-            wh_match = wh_match and (test_tags[i][0] == predicted_tags[i][0])
-            i += 1
-        return wh_match
+            # Add WH words with weight 60
+            elif self.is_wh_word(token):
+                # .lemma_ is already lowercase; no .lower() needed
+                features[token.lemma_] = 60
 
-    def classify_question(self, test_question):
-        """
-        Match a user query with a question in the database based on the classifier we trained and overall features we calculated.
-        Return relevant question.
-        """
+            # Add all other words with weight 30
+            else:
+                features[token.lemma_] = 30
+
+        # Replace the stemmed main verb with weight 60
+        sent = next(spacy_doc.sents)
+        stemmed_main_verb = sent.root.lemma_
+        features[stemmed_main_verb] = 60
+
+        return features
+
+    def classify_question(self, question):
         if self.classifier is None:
-            raise ValueError("Classifier not initialized")
-
-        #if self.use_new:
-        test_features = self.get_question_features(test_question)
-        #else:
-        #    test_features = self.get_question_features_old_algorithm(
-        #        test_question)
-        test_vector = dict.fromkeys(self.overall_features, 0)
-        for key in test_features:
-            if key in test_vector:
-                test_vector[key] = test_features[key]
-            #else:
-                # IF A WORD IS NOT IN THE EXISTING FEATURE SET, IT MAY BE A QUESTION WE CANNOT ANSWER.
-            #    test_vector["not related"] += 250
-        test_vector = np.array(list(test_vector.values()))
-        test_vector = test_vector.reshape(1, -1)
-        min_dist = np.min(self.classifier.kneighbors(test_vector, n_neighbors=1)[0])
+            raise ValueError("Classifier is not initialized")
+
+        # Create the spacy doc. Handles pos tagging, stop word removal, tokenization,
+        # lemmatization, etc
+        doc = self.nlp(question)
+        test_features = self.get_question_features(doc)
+
+        array_gen = [
+            test_features[k] if k in test_features else 0 for k in self.overall_features
+        ]
+        test_array = np.array(array_gen)
+
+        # Flatten array into a vector
+        test_vector = test_array.reshape(1, -1)
```

Comment:
Nothing against reshape here, but I'm curious as to whether there's some benefit of using _reshape()_ here over _flatten()_